### PR TITLE
[UNR-2600] Offloading & Authority & Startup actor fixes

### PR DESCRIPTION
**Pull request by mironec**
#### Description
Updates Matt's fixes for offloading and startup actors with latest changes from master. Unfortunately had to make a hard break between load balancing and offloading for now, as this change makes them incompatible for now.

4.22 Engine PR: https://github.com/improbableio/UnrealEngine/pull/337
4.23 Engine PR: https://github.com/improbableio/UnrealEngine/pull/338

#### Release note
- When running with Offloading enabled, Actors will have local authority (ROLE_Authority) on servers for longer periods of time to allow more native Unreal functionality to work without problems.
- When running with Offloading enabled, and trying to spawn Actors on a server which will not be the Actor Group owner for them, an error is logged and the Actor is deleted.

#### Tests
Tested with handover gym, cross-server-rpc gym, crashbot tutorial.

#### Documentation
In code

#### Primary reviewers
@mattyoung-improbable @alastairdglennie

### CODE REVIEW

**Comment by alastairdglennie** [comment]
Gonna conflict with my changes for making startup Actors work as desired with zoning, will come find you after lunch to ask questions, I can generally just rebase off you though

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 1]
```diff
@@ -794,6 +797,9 @@ void USpatialNetDriver::BeginDestroy()
 		GDKServices->GetLocalDeploymentManager()->OnDeploymentStart.Remove(SpatialDeploymentStartHandle);
 	}
 #endif
+
+	// null out reference to ActorGroupManger
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: comment adds nothing
```suggestion
```

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Removed comment

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp [hunk 2]
```diff
@@ -196,3 +202,69 @@ void USpatialGameInstance::HandleOnConnectionFailed(const FString& Reason)
...
+					ENetRole Temp = Actor->Role;
+					Actor->Role = Actor->RemoteRole;
+					Actor->RemoteRole = Temp;
+				}
+			}
+		}
+
+		if (GetDefault<USpatialGDKSettings>()->bEnableUnrealLoadBalancer)
+		{
+			check(!USpatialStatics::IsSpatialOffloadingEnabled());
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp]: I think we should prevent people from running with both enable rather than leaving this to runtime. Can we added something to `FSpatialGDKEditorToolbarModule::ValidateGeneratedLaunchConfig`?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 3]
```diff
@@ -600,6 +603,35 @@ void USpatialNetDriver::QueryGSMToLoadMap()
...
+
+	if (Actor->GetLocalRole() != ROLE_Authority)
+	{
+		return;
+	}
+
+	if (USpatialStatics::IsActorGroupOwnerForActor(Actor))
+	{
+		return;
+	}
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Is this a pattern we're using in the GDK? What's justification vs. just using once conditional with ORs?

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: I am happy to change it to ORs, but I do think we have this pattern elsewhere as well.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 4]
```diff
@@ -600,6 +603,35 @@ void USpatialNetDriver::QueryGSMToLoadMap()
...
+
+	if (Actor->GetLocalRole() != ROLE_Authority)
+	{
+		return;
+	}
+
+	if (USpatialStatics::IsActorGroupOwnerForActor(Actor))
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: If you're not using offloading, are is the default Actor group or all Actors just the UnrealWorker one?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp (hunk 2 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp]: I see, yes, that would be a good place, but will probably leave the checks here as well, to catch cloud deployments and general other flows.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 5]
```diff
@@ -600,6 +603,35 @@ void USpatialNetDriver::QueryGSMToLoadMap()
...
+	{
+		return;
+	}
+
+	FString WorkerType = GetGameInstance()->GetSpatialWorkerType().ToString();
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Seems weird to me we don't grab this by ref?
```suggestion
	const FString WorkerType = GetGameInstance()->GetSpatialWorkerType().ToString();
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 4 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Basically, yes, I think it's called the default actor group and the UnrealWorker is the owner of all of it.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp [hunk 6]
```diff
@@ -594,6 +595,11 @@ void UGlobalStateManager::BeginDestroy()
 
 void UGlobalStateManager::BecomeAuthoritativeOverAllActors()
 {
+	// This logic is not used in offloading.
+	if (USpatialStatics::IsSpatialOffloadingEnabled()) {
+		return;
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp]:
```suggestion
	if (USpatialStatics::IsSpatialOffloadingEnabled()) 
	{
		return;
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp [hunk 7]
```diff
@@ -64,11 +65,19 @@ TArray<FWorkerComponentData> EntityFactory::CreateEntityComponents(USpatialActor
...
 	const FClassInfo& Info = ClassInfoManager->GetOrCreateClassInfoByClass(Class);
-	WorkerAttributeSet WorkerAttributeOrSpecificWorker{ Info.WorkerType.ToString() };
+
+	FName EffectiveWorkerType = Info.WorkerType;
+	const USpatialGDKSettings* SpatialSettings = GetDefault<USpatialGDKSettings>();
+
+	if (SpatialSettings->bEnableOffloading)
+	{
+		EffectiveWorkerType = ActorGroupManager->GetWorkerTypeForActorGroup(USpatialStatics::GetActorGroupForActor(Actor));
+	}
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp]: Personal pref, change if you like:

```suggestion
        const USpatialGDKSettings* SpatialSettings = GetDefault<USpatialGDKSettings>();

	const FName EffectiveWorkerType = SpatialSettings->bEnableOffloading ? 
	         ActorGroupManager->GetWorkerTypeForActorGroup(USpatialStatics::GetActorGroupForActor(Actor)) :
	         Info.WorkerType;
```

Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp]: also prefer `EffectiveWorkerType` -> `AclAuthoritativeWorkerAttribute`

File: SpatialGDK/Source/SpatialGDK/Public/Utils/EntityFactory.h [hunk 8]
```diff
@@ -3,6 +3,8 @@
 #pragma once
  
 #include "SpatialCommonTypes.h"
+#include "Utils/SpatialActorGroupManager.h"
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/EntityFactory.h]: Can forward declare instead?

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h [hunk 9]
```diff
@@ -152,10 +152,11 @@ class SPATIALGDK_API USpatialNetDriver : public UIpNetDriver
 	UPROPERTY()
 	USpatialWorkerFlags* SpatialWorkerFlags;
 
-	TUniquePtr<SpatialActorGroupManager> ActorGroupManager;
 	TUniquePtr<SpatialLoadBalanceEnforcer> LoadBalanceEnforcer;
 	TUniquePtr<SpatialVirtualWorkerTranslator> VirtualWorkerTranslator;
 
+	SpatialActorGroupManager* ActorGroupManager;
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h]: Can we forward declare the import instead?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp (hunk 2 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp]: Fair enough, can we check somewhere earlier though, i.e. the init method here?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp [hunk 10]
```diff
@@ -196,3 +202,69 @@ void USpatialGameInstance::HandleOnConnectionFailed(const FString& Reason)
...
+				{
+					UE_LOG(LogSpatialGameInstance, Verbose, TEXT("WorkerType %s Not actor group owner of startup actor %s, Exchanging Roles"), *WorkerType, *GetPathNameSafe(Actor));
+					ENetRole Temp = Actor->Role;
+					Actor->Role = Actor->RemoteRole;
+					Actor->RemoteRole = Temp;
+				}
+			}
+		}
+
+		if (GetDefault<USpatialGDKSettings>()->bEnableUnrealLoadBalancer)
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp]:
```suggestion
		else if (GetDefault<USpatialGDKSettings>()->bEnableUnrealLoadBalancer)
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp (hunk 7 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp]: I like it, changed AclAuthoritativeWorkerAttribute to AclAuthoritativeWorkerType, as it's just a string at this point.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp]: well, an FName, but you know what I mean

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 11]
```diff
@@ -669,24 +691,19 @@ void USpatialNetDriver::OnLevelAddedToWorld(ULevel* LoadedLevel, UWorld* OwningW
...
-		// If load balancing is disabled and this worker is not GSM authoritative then exit early.
+		// If load balancing is disabled then exit early.
 		return;
 	}
 
 	for (auto Actor : LoadedLevel->Actors)
 	{
-		// If load balancing is disabled, we must be the GSM-authoritative worker, so set Role_Authority
-		// otherwise, load balancing is enabled, so check the lb strategy.
 		if (Actor->GetIsReplicated() &&
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]:
```suggestion
		if (Actor->GetIsReplicated() && LoadBalanceStrategy->ShouldHaveAuthority(*Actor))
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 12]
```diff
@@ -669,24 +691,19 @@ void USpatialNetDriver::OnLevelAddedToWorld(ULevel* LoadedLevel, UWorld* OwningW
...
-			(!bLoadBalancingEnabled || LoadBalanceStrategy->ShouldHaveAuthority(*Actor)))
+			(LoadBalanceStrategy->ShouldHaveAuthority(*Actor)))
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 13]
```diff
@@ -883,19 +883,22 @@ void USpatialReceiver::RemoveActor(Worker_EntityId EntityId)
 
 	if (USpatialActorChannel* ActorChannel = NetDriver->GetActorChannelByEntityId(EntityId))
 	{
-		for (UObject* SubObject : ActorChannel->CreateSubObjects)
+		if (NetDriver->GetWorld() != nullptr && !NetDriver->GetWorld()->IsPendingKillOrUnreachable())
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: This is a drive-by fix to a crash when exiting PIE, where the world context could be null. @ImprobableNic

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: (Seems to only happen with multiple servers)

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp [hunk 14]
```diff
@@ -195,3 +203,56 @@ void USpatialGameInstance::HandleOnConnectionFailed(const FString& Reason)
...
+				else
+				{
+					UE_LOG(LogSpatialGameInstance, Verbose, TEXT("WorkerType %s is not the actor group owner of startup actor %s, exchanging Roles"), *WorkerType, *GetPathNameSafe(Actor));
+					ENetRole Temp = Actor->Role;
+					Actor->Role = Actor->RemoteRole;
+					Actor->RemoteRole = Temp;
+				}
+			}
+		}
+
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 15]
```diff
@@ -600,6 +604,24 @@ void USpatialNetDriver::QueryGSMToLoadMap()
 	GlobalStateManager->QueryGSM(QueryDelegate);
 }
 
+void USpatialNetDriver::OnActorSpawned(AActor* Actor)
+{
+	if (!Actor->GetIsReplicated() ||
+		Actor->GetLocalRole() != ROLE_Authority ||
+		Actor->GetClass()->HasAnySpatialClassFlags(SPATIALCLASS_Singleton | SPATIALCLASS_NotSpatialType) ||
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: stupid singletons

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 16]
```diff
@@ -600,6 +604,24 @@ void USpatialNetDriver::QueryGSMToLoadMap()
...
+		USpatialStatics::IsActorGroupOwnerForActor(Actor))
+	{
+		// We only want to delete actors which are replicated and we somehow gain local authority over, while not the actor group owner
+		return;
+	}
+
+	FString WorkerType = GetGameInstance()->GetSpatialWorkerType().ToString();
+	UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Worker %s spawned replicated actor %s (owner: %s) but is not actor group owner for actor group %s. The actor will be destroyed in 0.01s"),
+		*WorkerType, *GetNameSafe(Actor), *GetNameSafe(Actor->GetOwner()), *USpatialStatics::GetActorGroupForActor(Actor).ToString());
+	Actor->TearOff();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Any reason you went for this instead of setting bReplicated to false?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp [hunk 17]
```diff
@@ -594,6 +595,12 @@ void UGlobalStateManager::BeginDestroy()
 
 void UGlobalStateManager::BecomeAuthoritativeOverAllActors()
 {
+	// This logic is not used in offloading.
+	if (USpatialStatics::IsSpatialOffloadingEnabled())
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp]: Add to `BecomeAuthoritativeOverActorsBasedOnLBStrategy` also?

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialGameInstance.h [hunk 18]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) Improbable Worlds Ltd, All Rights Reserved
 
 #pragma once
+#include "Utils/SpatialActorGroupManager.h"
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialGameInstance.h]:
```suggestion

#include "Utils/SpatialActorGroupManager.h"
```

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h [hunk 19]
```diff
@@ -41,6 +40,7 @@ class USpatialSender;
 class USpatialStaticComponentView;
 class USpatialWorkerConnection;
 class USpatialWorkerFlags;
+class SpatialActorGroupManager;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h]: Sort please

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialStatics.cpp [hunk 20]
```diff
@@ -129,8 +135,13 @@ FName USpatialStatics::GetActorGroupForActor(const AActor* Actor)
 {
 	if (SpatialActorGroupManager* ActorGroupManager = GetActorGroupManager(Actor))
 	{
-		UClass* ActorClass = Actor->GetClass();
-		return ActorGroupManager->GetActorGroupForClass(ActorClass);
+		const AActor* EffectiveActor = Actor;
+		while (EffectiveActor->bUseNetOwnerActorGroup && EffectiveActor->GetOwner() != nullptr)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialStatics.cpp]: Use this same logic in `IsActorGroupOwnerForActor` ?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp (hunk 2 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp]: Moved to init method

File: SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp (hunk 17 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp]: `BecomeAuthoritativeOverActorsBasedOnLBStrategy` is only called when the Unreal Load Balancer is enabled, so offloading cannot be enabled at the same time.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h (hunk 19 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h]: Sorted, but not sure if well. Since everything here begins with U because they are UObjects, this begins with S, so goes above all of them.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp (hunk 17 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp]: Defensive coding!

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp [hunk 21]
```diff
@@ -195,3 +203,55 @@ void USpatialGameInstance::HandleOnConnectionFailed(const FString& Reason)
...
+					Actor->RemoteRole = Temp;
+				}
+			}
+		}
+		else if (GetDefault<USpatialGDKSettings>()->bEnableUnrealLoadBalancer)
+		{
+			if (Actor->GetIsReplicated())
+			{
+				// Always wait for authority to be delegated down from SpatialOS, if using zoning
+				Actor->Role = ROLE_SimulatedProxy;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp]: We're only doing this now if the load balancer is enabled - which wasn't the case previously. Become authoritative over all actors should get called in either instance, so possibly safe to keep behaviour consistent.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 16 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Spoke offline, added a comment explaining this

**Comment by mironec** [comment]
Added the discussed changes: https://github.com/spatialos/UnrealGDK/pull/1762/commits/eec686e100baae5d6c21cefccf7ee9ddc43b5c01
@mattyoung-improbable @m-samiec

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h [hunk 22]
```diff
@@ -136,6 +137,11 @@ class SPATIALGDK_API USpatialActorChannel : public UActorChannel
 
 		if (EntityId != SpatialConstants::INVALID_ENTITY_ID)
 		{
+			// Make sure we have spatial authority before we replicate with Offloading, as we pretend to have local authority
+			if (USpatialStatics::IsSpatialOffloadingEnabled() && !NetDriver->StaticComponentView->HasAuthority(EntityId, SpatialConstants::POSITION_COMPONENT_ID)) {
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h]:
```suggestion
			if (USpatialStatics::IsSpatialOffloadingEnabled() && !NetDriver->StaticComponentView->HasAuthority(EntityId, SpatialConstants::POSITION_COMPONENT_ID))
			{
```

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h [hunk 23]
```diff
@@ -136,6 +137,11 @@ class SPATIALGDK_API USpatialActorChannel : public UActorChannel
 
 		if (EntityId != SpatialConstants::INVALID_ENTITY_ID)
 		{
+			// Make sure we have spatial authority before we replicate with Offloading, as we pretend to have local authority
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h]: Update comment

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h [hunk 24]
```diff
@@ -136,6 +137,12 @@ class SPATIALGDK_API USpatialActorChannel : public UActorChannel
...
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h]:
```suggestion
			// If the entity already exists, make sure we have spatial authority before we replicate with Offloading, because we pretend to have local authority
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 25]
```diff
@@ -600,6 +604,25 @@ void USpatialNetDriver::QueryGSMToLoadMap()
 	GlobalStateManager->QueryGSM(QueryDelegate);
 }
 
+void USpatialNetDriver::OnActorSpawned(AActor* Actor)
+{
+	if (!Actor->GetIsReplicated() ||
+		Actor->GetLocalRole() != ROLE_Authority ||
+		Actor->GetClass()->HasAnySpatialClassFlags(SPATIALCLASS_Singleton | SPATIALCLASS_NotSpatialType) ||
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Update to !SpatialType

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Changed to 
```Actor->GetClass()->HasAnySpatialClassFlags(SPATIALCLASS_Singleton) ||
!Actor->GetClass()->HasAnySpatialClassFlags(SPATIALCLASS_SpatialType) ||
```

**Comment by mironec** [comment]
Updated with the latest suggestions, added changelog notes.

**Comment by mironec** [comment]
4.24 Engine PR: https://github.com/improbableio/UnrealEngine/pull/391

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h [hunk 26]
```diff
@@ -136,6 +137,12 @@ class SPATIALGDK_API USpatialActorChannel : public UActorChannel
 
 		if (EntityId != SpatialConstants::INVALID_ENTITY_ID)
 		{
+			// If the entity already exists, make sure we have spatial authority before we replicate with Offloading, because we pretend to have local authority
+			if (USpatialStatics::IsSpatialOffloadingEnabled() && !NetDriver->StaticComponentView->HasAuthority(EntityId, SpatialConstants::POSITION_COMPONENT_ID))
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h]: I would like to use something else than the position component, but I don't think we have a nice function for checking authority over the actually relevant components.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 27]
```diff
@@ -703,8 +715,8 @@ void USpatialNetDriver::OnLevelAddedToWorld(ULevel* LoadedLevel, UWorld* OwningW
 		if (Actor->GetIsReplicated() &&
 			(!bLoadBalancingEnabled || LoadBalanceStrategy->ShouldHaveAuthority(*Actor)))
 		{
-			Actor->Role = ROLE_Authority;
-			Actor->RemoteRole = ROLE_SimulatedProxy;
+				Actor->Role = ROLE_Authority;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Is this double tabbed?

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Yes, it was, nice catch, made single tab.
